R/derive_vars_cat.R

```diff
@@ -0,0 +1,256 @@
+#' Derive Categorization Variables Like `AVALCATy` and `AVALCAyN`
+#' @param dataset
+#' `r roxygen_param_dataset(expected_vars = c("by_vars", "definition"))`
+#' @param definition List of expressions created by `exprs()`.
+#' Must be in rectangular format and specified using the same syntax as when creating
+#' a `tibble` using the `tribble()` function.
+#' The `definition` object it will be converted to a `tibble` using the `tribble()` function.
+#'
+#' Must contain:
+#'  - the column `condition` which evaluates to a logic in `dataset`.
+#'  - at least one additional column with the new column name and the category value.
+#'  - the column specified in `by_vars` (if `by_vars` is specified)
+#'
+#' e.g. if `by_vars` is not specified:
+#'
+#' ```{r}
+#' #| eval: false
+#' exprs(~condition,   ~AVALCAT1, ~AVALCA1N,
+#'       AVAL >= 140, ">=140 cm",         1,
+#'       AVAL < 140,   "<140 cm",         2)
+#' ```
+#'
+#' e.g. if `by_vars` is specified as `exprs(VSTEST)`:
+#'
+#' ```{r}
+#' #| eval: false
+#' exprs(~VSTEST,   ~condition,  ~AVALCAT1, ~AVALCA1N,
+#'       "Height", AVAL >= 140, ">=140 cm",         1,
+#'       "Height",  AVAL < 140,  "<140 cm",         2)
+#' ```
+#'
+#' @param by_vars list of expressions with one element. `NULL` by default.
+#' Allows for specifying by groups, e.g. `exprs(PARAMCD)`.
+#' Variable must be present in both `dataset` and `definition`.
+#' The conditions in `definition` are applied only to those records that match `by_vars`.
+#' The categorization variables are set to NA for records
+#' not matching any of the by groups in `definition`.
+#'
+#'
+#' @details
```

Comment:
Not for this PR - but just thinking should we alert the user that something is amiss with their logic if  they don't abide by our warning?  Not sure if we can detect this first??  @bundfussr

Reply:
@bms63 , do you mean issuing a warning if there are records which doesn't match any of the by groups?

Reply:
In the overlap discussion that is in the documentation - like something should be finer in logic??

Reply:
I'm not sure if this would be helpful. There are use cases where it makes sense to use overlapping categories, e.g.,
```
~condition,    ~ANRIND,
AVAL > ANRHI,  "high",
AVAL <= ANRHI, "normal",
TRUE,          "unknown"
```
Issuing a warning in this case would force the users to suppress the warning or replace `TRUE` with `is.na(AVAL) | is.na(ANRHI)`.